### ISSUE TITLE
Issues identified with uncaught throttling errors

### ISSUE THREAD

**Issue by tloubrieu-jpl** [NASA-PDS/harvest#206]
### Checked for duplicates

No - I haven't checked

### 🐛 Describe the bug

ATM node (Irma Trejo), is having issues with harvest.

See logs:
 
```
2024-10-25 14:27:17,938 [ERROR] LIDVID = urn:nasa:pds:juno_mwr:data_calibrated:mwr14rg2018175130000_r04112_v03::2.1, Message = rejected execution of primary operation [throttled]
2024-10-25 14:27:17,939 [ERROR] LIDVID = urn:nasa:pds:juno_mwr:data_calibrated:mwr14rg2018176160000_r04112_v03::2.1, Message = rejected execution of primary operation [throttled]
2024-10-25 14:27:17,939 [ERROR] LIDVID = urn:nasa:pds:juno_mwr:data_calibrated:mwr14rg2018176040000_r04112_v03::2.1, Message = rejected execution of primary operation [throttled]
2024-10-25 14:27:17,939 [ERROR] LIDVID = urn:nasa:pds:juno_mwr:data_calibrated:mwr14rg2018176050000_r04120_v03::2.1, Message = rejected execution of primary operation [throttled]
```

And finally:

`2024-10-25 15:51:14,619 [ERROR] Unknown host b3rqys09xmx9i19yn64i.us-west-2.aoss.amazonaws.com
` 

It is likely that it comes from an AWS opensearch server instability, so we want to do, possibly 2 things:
- reduce this instability
- make harvest robust to these events.

### 🕵️ Expected behavior

It is likely that it comes from an AWS opensearch server instability, so we want to do, possibly, 2 things:
- reduce this instability
- make harvest robust to these events.

### 📜 To Reproduce

1.
2.
3.
...

### 🖥 Environment Info

- Version of this software [e.g. vX.Y.Z]
- Operating System: [e.g. MacOSX with Docker Desktop vX.Y]
...

### 📚 Version of Software Used

harvest version 4.0.2

### 🦄 Related requirements

🦄 #xyz

**Comment by tloubrieu-jpl** [comment]
What we can look at first:
- stress test one opensearch index to reproduce the thrttling limit and get statistics on it
- look at the opensearch api response in case they contain anything related to the API limits
- look at opensearchpy if there is a built-in throttling management.

**Comment by tloubrieu-jpl** [comment]
@alexdunnjpl @al-niessner , I found this elastic search documentation which provides information on the api response header which can be used to pace the client requests. See https://www.elastic.co/guide/en/cloud/current/ec-api-rate-limiting.html

I did not have a chance to test the availability of this headers on our opensearch server yet.

**Comment by alexdunnjpl** [comment]
@tloubrieu-jpl @jordanpadams @sjoshi-jpl with the `opensearch-py` library I'm getting the following back at the point where AOSS throttles:

```JSON
{'update': {'_index': 'atm-registry', '_id': 'urn:nasa:pds:lab.hydrocarbon_spectra::1.1', 'status': 429, 'error': {'type': 'circuit_breaking_exception', 'reason': 'rejected execution of primary operation [throttled]', 'bytes_wanted': 0, 'bytes_limit': 0, 'durability': 'TRANSIENT'}}}
```

I had hoped to use the AOSS dev console to make a raw request while under throttling, but it looks like our opensearch dev user doesn't actually have the necessary permissions to make doc updates?

```
POST /atm-registry/_bulk
{"update": {"_id": "urn:nasa:pds:lab.hydrocarbon_spectra::1.1"}}
{"doc": {"ops:Provenance/ops:reindexed_at": "2024-10-30T17:01:30.881195"}}
{"update": {"_id": "urn:nasa:pds:lab.hydrocarbon_spectra:context::1.1"}}
{"doc": {"ops:Provenance/ops:reindexed_at": "2024-10-30T17:01:30.881195"}}
...
```

```JSON
{
  "took": 53,
  "errors": true,
  "items": [
    {
      "update": {
        "_index": "atm-registry",
        "_id": "urn:nasa:pds:lab.hydrocarbon_spectra::1.1",
        "status": 403,
        "error": {
          "type": "security_exception",
          "reason": "Authorization failure for the following indices: [atm-registry]"
        }
      }
    },
    {
      "update": {
        "_index": "atm-registry",
        "_id": "urn:nasa:pds:lab.hydrocarbon_spectra:context::1.1",
        "status": 403,
        "error": {
          "type": "security_exception",
          "reason": "Authorization failure for the following indices: [atm-registry]"
        }
      }
    },
    ...
  ]
}
```

@jordanpadams @tloubrieu-jpl how much time do you want me to spend on sorting this ASAP (figuring out auth for the raw requests and making the requests from prod MCP EC2), vs waiting until Sagar's back and getting the necessary permissions to test this using the dev console?

**Comment by jordanpadams** [comment]
@viviant100 can you help @alexdunnjpl with above?

**Comment by viviant100** [comment]
> @viviant100 can you help @alexdunnjpl with above?
Temporary permissions have been set up for MCP prod testing and will be reverted once testing is complete.

**Comment by alexdunnjpl** [comment]
@tloubrieu-jpl @jordanpadams some results from my testing:

- The errors, this time, are very transient, to the extent that it was difficult to send a single request manually in a way that it also returned a 429.  This suggests that retrying is _not_ a problematic strategy.

- The errors in question, by virtue of being circuit-breaker exceptions - are the result of the server detecting an imminent out-of-memory condition and rejecting search or index operations on that basis.  They are not the result of any explicit limit, ergo:

  - the errors are universal (presumably to the cluster, since the cluster shares resources), and cannot be configured or isolated per user.

  - the server is unable to indicate when to back off or how long to wait before retrying, as that information is not available (because there is no rate limit/period).  I haven't managed to access the headers for a 429 response, but would suggest that doing so would be a waste of time/effort.

My suggestion would be for any tool (sweepers, harvest, etc) which has the potential to heavily load index or search, to implement 429-specific retry functionality in addition to any other "log-and-continue" retry functionality already present (for example, sweepers will log and continue to prevent a missing product from blocking the 99.999% of non-problematic products.

I'll implement this in sweepers asap.

**Comment by alexdunnjpl** [comment]
@viviant100 @sjoshi-jpl I've reverted the temporary grant of admin permissions to the dev role, as requested by @viviant100

**Comment by viviant100** [comment]
Thanks @alexdunnjpl !

**Comment by al-niessner** [comment]
@alexdunnjpl @tloubrieu-jpl 

Do a 10-40 second random delay? I can add this to harvest but if all tools do same random delay it may help reduce quick repeats of the failure. It mimics the ethernet protocol which has similar collision issues. Trying again as fast as possible will generate the same failure quickly.

**Comment by alexdunnjpl** [comment]
@al-niessner @tloubrieu-jpl the issue, as I understand it, is sustained burst load, generally from a single source or a small number of clients (for harvest and the write-heavy sweepers, at least).

Adding random jitter is intended to avoid the case where many clients fail at the same time and "coincidentally" retry at the same time, which is not the case here.  That said, it certainly wouldn't cause any harm.  But it *is* important imho that the backoff contain a judicious expo component, such that repeated failures indicate an overwhelmed AOSS and result in the client(s) giving it time to process its backlog and desaturate its compute resources.

**Comment by al-niessner** [comment]
@alexdunnjpl 

I am good with that 1-4 seconds then 10-40 seconds then 100-400 seconds then quit?

**Comment by tloubrieu-jpl** [comment]
@al-niessner , @alexdunnjpl before you move forward, can you check if you have access to the header variables described there https://www.elastic.co/guide/en/cloud/current/ec-api-rate-limiting.html in the opensearch api responses.

Sorry I thought I posted that yesterday.

**Comment by alexdunnjpl** [comment]
@tloubrieu-jpl you did (in another ticket, I think)

Did you read my observations here?

https://github.com/NASA-PDS/harvest/issues/206#issuecomment-2448629509

I can check, if you insist, but it will be time-consuming and I'm fairly confident the answer will be no, for the reasons given.

**Comment by alexdunnjpl** [comment]
> @alexdunnjpl
> 
> I am good with that 1-4 seconds then 10-40 seconds then 100-400 seconds then quit?

Personally I'd tend for less-aggressive expo, maybe a factor of 2-4 rather than 10, but that's your call.

**Comment by tloubrieu-jpl** [comment]
Ok @alexdunnjpl , the reason for investigating the header response is also to better understand how to properly manage throttling as a user but also as a provider. We can keep this analysis for later. In a different ticket with lower priority.

**Comment by alexdunnjpl** [comment]
@tloubrieu-jpl sounds good!

**Comment by c-suh** [comment]
@tloubrieu-jpl, @al-niessner, and @alexdunnjpl, please provide the instructions and resources to test this. Thank you.

**Comment by c-suh** [comment]
i&t.skip label added because it is not worth reproducing a heavy load in production from registry.
